maintainers_annotation/maintainers_annotation.go

```diff
@@ -0,0 +1,190 @@
+)
+
+var errNoScorecardYmlFile = errors.New("scorecard.yml doesn't exist or file doesn't contain exemptions")
+
+// AnnotationReason is the reason behind an annotation.
+type AnnotationReason string
+
+const (
+	// AnnotationTestData is to annotate when a check or probe is targeting a danger
+	// in files or code snippets only used for test or example purposes.
+	AnnotationTestData AnnotationReason = "test-data"
+	// AnnotationRemediated is to annotate when a check or probe correctly identified a
+	// danger and, even though the danger is necessary, a remediation was already applied.
+	// E.g. a workflow is dangerous but only run under maintainers verification and approval,
+	// or a binary is needed but it is signed or has provenance.
+	AnnotationRemediated AnnotationReason = "remediated"
+	// AnnotationNotApplicable is to annotate when a check or probe is not applicable for the case.
+	// E.g. the dependencies should not be pinned because the project is a library.
+	AnnotationNotApplicable AnnotationReason = "not-applicable"
+	// AnnotationNotSupported is to annotate when the maintainer fulfills a check or probe in a way
+	// that is not supported by Scorecard. E.g. Clang-Tidy is used as SAST tool but not identified
+	// because its not supported.
+	AnnotationNotSupported AnnotationReason = "not-supported"
+	// AnnotationNotDetected is to annotate when the maintainer fulfills a check or probe in a way
+	// that is supported by Scorecard but not identified. E.g. Dependabot is configured in the
+	// repository settings and not in a file.
+	AnnotationNotDetected AnnotationReason = "not-detected"
+)
+
+// AnnotationReasonDoc maps the reasons to their human-readable explanations.
+var AnnotationReasonDoc = map[AnnotationReason]string{
+	AnnotationTestData:      "The files or code snippets are only used for test or example purposes.",
+	AnnotationRemediated:    "The dangerous files or code snippets are necessary but remediations were already applied.",
+	AnnotationNotApplicable: "The check or probe is not applicable in this case.",
+	AnnotationNotSupported:  "The check or probe is fulfilled but in a way that is not supported by Scorecard.",
+	AnnotationNotDetected:   "The check or probe is fulfilled but in a way that is supported by Scorecard but it was not detected.",
+}
+
+// Annotation groups the annotation reason and, in the future, the related probe.
+// If there is a probe, the reason applies to the probe.
+// If there is not a probe, the reason applies to the check or group of checks in
+// the exemption.
+type Annotation struct {
+	AnnotationReason AnnotationReason `yaml:"annotation"`
+}
+
+// Exemption defines a group of checks that are being annotated for various reasons.
+type Exemption struct {
+	Checks      []string     `yaml:"checks"`
+	Annotations []Annotation `yaml:"annotations"`
+}
+
+// MaintainersAnnotation is a group of exemptions.
+type MaintainersAnnotation struct {
+	Exemptions []Exemption `yaml:"exemptions"`
+}
```

Comment:
Actually, if the idea is to add more configs (beyond annotating checks) to the .scorecard.yml file in the future*, might it make sense to move all the scorecard.yml reading and parsing to a separate module?

```suggestion
type ScorecardConfig struct {
	MaintainerAnnotations []Exemption `yaml:"exemptions"`
}
```

```go
// config/config.go
import "maintainer_annotations"

func parseScorecardConfigFile(cfg *ScorecardConfig, content[]byte) error {...}
func readScorecardConfigFromRepo(...) (ScorecardConfig, error) {...}
func GetScorecardConfig(...) (ScorecardConfig, error) {...}
```

Whenever the maintainer annotations are needed, they're in scorecardConfig.MaintainerAnnotations.

The maintainer_annotation/ folder would contain all the annotation-specific logic (`IsExempted`, `getAnnotations`).

---

This may be overkill at this very moment (since the ScorecardConfig === MaintainerAnnotations), but if the idea is to add more configs in the future, moving these functions elsewhere later would be a breaking change for any other projects using those functions, like this one:

```go
// github.com/pnacht/scorecard_super_scraper
import ".../scorecard/maintainer_annotation"

func scrapeAllTheReposAndTheirScorecardStuffs() {
    for repo := range GetAllTheRepos() {
        // ... a lot of scraping ...
        GetMaintainerAnnotations()  // breaks if no longer in /maintainer_annotation
        // ...
    }
}
```
---

\* As I'd understood from the design doc and the choice of keeping the annotations within their own `exemptions:` field

Reply:
@spencerschrock @raghavkaul Any comments on this renaming changes?

Reply:
On my end, LGTM to all that. I do prefer that scorecard.yml format, too (I prefer `annotations:` over `exemptions:`, given the feature is called Maintainer Annotations).

Reply:
+1 to Pedro's point of handling the file as a generic settings file so we dont break people later.

Naming is very subjective. I lean towards a simple "config" package.
`config/config.go` would have a `Parse` function, so callers would do `c := config.Parse(foo)`

Depending on how complex the annotation specific logic is, I could see it in `config/annotations.go` or `config/annotations/annotations.go`

Adding `Scorecard` to the various names is redundant in my opinion.

Reply:
Please review the renamings @pnacht